README.md

```diff
@@ -219,54 +367,171 @@ for (padded_string json : { string("[1, 2, 3]"), string("true"), string("[ true,
+set architecture-specific flags yourself (e.g., `-mavx2`, `/AVX2`  or `-march=haswell`), and it may
+even break runtime dispatch and your binaries will fail to run on older processors.
+
+#### Runtime CPU Detection
+
+When you first use simdjson, it will detect the CPU you're running on, and swap over to the fastest
+implementation for it. This is a small, one-time cost and for many people will be paid the first
+time they call `parse()` or `load()`.
+
+You can check what implementation is running with `active_implementation`:
 
 ```c++
-document::parser parser(0); // This parser is not allowed to auto-allocate
-auto alloc_error = parser.set_capacity(1024*1024); // Set initial capacity to 1MB
-if (alloc_error) { exit(1); };
+cout << "simdjson v" << #SIMDJSON_VERSION << endl;
+cout << "Detected the best implementation for your machine: " << simdjson::active_implementation->name();
+cout << "(" << simdjson::active_implementation->description() << ")" << endl;
+```
 
-for (int i=0;i<argc;i++) {
-  auto [doc, error] = parser.parse(get_corpus(argv[i]));
-  if (error == CAPACITY) { cerr << "JSON files larger than 1MB are not supported!" << endl; exit(1); }
-  if (error) { cerr << error << endl; exit(1); }
-  cout << doc;
+Implementation detection will happen in this case when you first call `name()`.
+
+#### Querying Available Implementations
+
+You can list all available implementations:
+
+```c++
+for (auto implementation : simdjson::available_implementations) {
+  cout << implementation->name() << ": " << implementation->description() << endl;
 }
 ```
 
-## Newline-Delimited JSON (ndjson) and  JSON lines
+And look them up by name:
 
-The simdjson library also support multithreaded JSON streaming through a large file containing many smaller JSON documents in either [ndjson](http://ndjson.org) or [JSON lines](http://jsonlines.org) format. If your JSON documents all contain arrays or objects, we even support direct file concatenation without whitespace. The concatenated file has no size restrictions (including larger than 4GB), though each individual document must be less than 4GB.
+```c++
+cout << simdjson::available_implementations["fallback"]->description() << endl;
+```
 
-Here is a simple example, using single header simdjson:
+#### Manually Selecting the Parser Implementation
 
-```cpp
-#include "simdjson.h"
-#include "simdjson.cpp"
+If you're trying to do performance tests or see how different implementations of simdjson run, you
+can select the CPU architecture yourself:
 
-int parse_file(const char *filename) {
-    simdjson::document::parser parser;
-    for (const document &doc : parser.load_many(filename)) {
-      // do something with the document ...
-    }
-}
+```c++
+// Use the fallback implementation, even though my machine is fast enough for anything
+simdjson::active_implementation = simdjson::available_implementations["fallback"];
 ```
 
-## Usage: easy single-header version
+### Large files and huge page support
+
```

Comment:
It seems that the section "Large files and huge page support" (which should probably be just "Large files") has a strong overlap with "Server Loops: Long-Running Processes and Memory Capacity" and "Reusing the parser for maximum efficiency". Can we try to merge these sections somehow? We don't need to keep a section called Large files per se...

Reply:
I kept it as a separate section, but removed the bits about reusing the parser (except for one sentence). I also moved it up closer to the sections recommending parser reuse.

Reply:
(I think huge page support is different enough from the other cases, and weird enough, to warrant its own explanation.)